generators/beagle/src/main/scala/BeaglePeriphery.scala

```diff
@@ -192,42 +265,141 @@ trait HasPeripheryBeagleBundle {
   }
 }
 
-trait HasPeripheryBeagleModuleImp extends LazyModuleImp with HasPeripheryBeagleBundle {
+trait HasPeripheryBeagleModuleImp extends LazyModuleImp with HasPeripheryBeagleBundle
+{
   val outer: HasPeripheryBeagle
 
+  // instantiate IOs
   val rst_async = IO(Input(Bool()))
   val boot = IO(Input(Bool()))
 
   val hbwif_clks = IO(Input(Vec(p(HbwifTLKey).numBanks, Clock()))) // clks for hbwif
   val hbwif_tx = IO(chiselTypeOf(outer.hbwif.module.tx))
   val hbwif_rx = IO(chiselTypeOf(outer.hbwif.module.rx))
 
-  // setup lbwif
   val lbwif_serial = IO(chiselTypeOf(outer.lbwif.module.io.ser))
-  lbwif_serial <> outer.lbwif.module.io.ser
 
-  val lbwifClkDiv = Module(new testchipip.ClockDivider(outer.scrParams.lbwifDividerBits))
-  val lbwif_clk = lbwifClkDiv.io.clockOut
-  val lbwif_rst = ResetCatchAndSync(lbwif_clk, reset.toBool)
-  lbwifClkDiv.io.divisor := outer.scr.module.io.lbwif_divider
+  // clocks coming from offchip
+  val single_clks = IO(Input(Vec(1, Clock())))
+  val diff_clks   = IO(Input(Vec(1, Clock())))
+
+  val bh_clk_sel     = IO(Input(Bool()))
+  val rs_clk_sel     = IO(Input(Bool()))
+  val uncore_clk_sel = IO(Input(Bool()))
+
+  val uncore_clk_out = IO(Output(Clock()))
+  val bh_clk_out     = IO(Output(Clock()))
+  val rs_clk_out     = IO(Output(Clock()))
+  val lbwif_clk_out  = IO(Output(Clock()))
+
+
+  // ------------------------------------------------
+
+  val scr_mod = outer.scr.module
+
+  // other signals
+  scr_mod.io.boot := boot
+  scr_mod.io.rst_async := rst_async
+
+  // ------------------------------------------------
+
+  // get the actual clock from the io clocks
+  require(bh_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(rs_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(uncore_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+
+  // setup uncore muxes and divider
+
+  val uncore_clk_mux = testchipip.ClockMutexMux(single_clks ++ diff_clks)
+  uncore_clk_mux.io.sel := uncore_clk_sel
+  uncore_clk_mux.io.resetAsync := rst_async
+  val uncore_clk_mux_out = uncore_clk_mux.io.clockOut
+  val uncore_clk_mux_rst = ResetCatchAndSync(uncore_clk_mux.io.clockOut, reset.asBool)
```

Comment:
How is `reset.asBool` different from `rst_async`?

Reply:
Isn't the normal `reset` the reset that the periphery is set to... aka it is the uncore reset. Then the `rst_async` is the reset gotten from offchip.